lib/os/queued_operation.c

```diff
@@ -0,0 +1,197 @@
+
+	return 0;
+}
+
+static inline bool can_start(/*const*/ struct queued_operation_manager *mgr)
+{
+	/* OK if not busy, not finalizing, and an op is available. */
+	return (mgr->current == NULL)
+	       && !mgr->finalizing
+	       && !sys_slist_is_empty(&mgr->operations);
+}
+
+static void select_next_and_unlock(struct queued_operation_manager *mgr,
+				   k_spinlock_key_t key)
+{
+	struct queued_operation *op = NULL;
+
+	/* Track whether the manager is idle, so we only send
+	 * notification of entry to idle once.
+	 */
+	bool in_idle = (mgr->current == NULL);
+
+	do {
+		/* Can't select a new operation if one is active. */
+		if (mgr->current != NULL) {
+			break;
+		}
+
+		/* Can't select a new operation until the previous one
+		 * completes finalization, lest we pick something that
+		 * has too low a priority.
+		 */
+		if (mgr->finalizing) {
+			break;
+		}
+
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->current = op;
+		} else {
+			op = NULL;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+
+		/* Only notify the manager if there's an operation, or
+		 * if it is to transition to idle.
+		 */
+		if ((op != NULL) || !in_idle) {
+			mgr->vtable->process(mgr, op);
+		}
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
+
+	k_spin_unlock(&mgr->lock, key);
+}
+
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority)
+{
+	int rv = 0;
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(mgr->vtable != NULL);
+	__ASSERT_NO_MSG(mgr->vtable->process != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	/* Validation is optional; if present, use it. */
+	if (mgr->vtable->validate) {
+		rv = mgr->vtable->validate(mgr, op);
+	}
+
+	/* Set the priority, checking whether it's in range. */
+	if (rv == 0) {
+		rv = op_set_priority(op, priority);
+	}
+
+	/* Reject callback notifications without translation
+	 * function.
+	 */
+	if ((rv == 0)
+	    && async_notify_uses_callback(&op->notify)
+	    && (mgr->vtable->callback == NULL)) {
+		rv = -ENOTSUP;
+	}
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	sys_slist_t *list = &mgr->operations;
+	struct queued_operation *prev = NULL;
+	struct queued_operation *tmp;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(list, tmp, node) {
+		if (priority < op_get_priority(tmp)) {
+			break;
+		}
+		prev = tmp;
+	}
+
+	if (prev == NULL) {
+		sys_slist_prepend(list, &op->node);
+	} else {
+		sys_slist_insert(list, &prev->node, &op->node);
+	}
+
+	select_next_and_unlock(mgr, key);
+
+	return rv;
+}
+
+void queued_operation_finalize(struct queued_operation_manager *mgr,
+			       struct queued_operation *op,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	if (mgr->current == op) {
+		mgr->finalizing = true;
```

Comment:
what is the purpose of `finalizing` flag? Is it to allow submitting new operation with higher priority from the callback?

Reply:
Yes, it ensures that any changes to the queued operations performed during the callback affect the selection of the next operation to run.